c10/cuda/CUDADeviceAssertion.h

```diff
@@ -0,0 +1,105 @@
+  return atomicAdd(&(assertions_data->assertion_count), 1);
+}
+
+__device__ __noinline__ void assertions_data_insert(
+    AssertionsData* assertions_data,
+    const char* assertion_msg0,
+    const char* filename0,
+    const char* function_name0,
+    const int line_number0,
+    const uint32_t caller0,
+    const dim3 block_id,
+    const dim3 thread_id) {
+  // `assertions_data` may be nullptr if device-side assertion checking
+  // is disabled at run-time. If it is disabled at compile time this
+  // function will never be called
+  if (!assertions_data) {
+    return;
+  }
+
+  const auto nid = next_assertions_id(assertions_data);
+  if (nid > C10_CUDA_DEVICE_SIDE_ASSERTION_COUNT) {
+    // At this point we're ran out of assertion buffer space
+    // we could print a message about this, but that'd get
+    // spammy if a lot of threads did it, so we just silently
+    // ignore any other assertion failures. In most cases the
+    // failures will all probably analogous anyway.
+    return;
+  }
+  auto& self = assertions_data->assertions[nid];
+  dstrcpy(self.assertion_msg, assertion_msg0);
+  dstrcpy(self.filename, filename0);
+  dstrcpy(self.function_name, function_name0);
+  self.line_number = line_number0;
+  self.caller = caller0;
+  self.block_id[0] = block_id.x;
+  self.block_id[1] = block_id.y;
+  self.block_id[2] = block_id.z;
+  self.thread_id[0] = thread_id.x;
+  self.thread_id[1] = thread_id.y;
+  self.thread_id[2] = thread_id.z;
+}
+
+// Emulates a kernel assertion. The assertion won't stop the kernel's progress,
+// so you should assume everything the kernel produces is garbage if there's an
+// assertion failure. NOTE: This assumes that `assertions_data` and `caller` are
+// arguments of the kernel and therefore accessible.
+#define CUDA_KERNEL_ASSERT2(condition)                                      \
+  do {                                                                      \
+    if (C10_UNLIKELY(!(condition))) {                                       \
+      /* Atomically increment so other threads can fail at the same time */ \
+      c10::cuda::assertions_data_insert(                                    \
+          assertions_data,                                                  \
+          C10_CUDA_DEVICE_ASSERTION_TOSTRING(condition),                    \
+          __FILE__,                                                         \
+          __FUNCTION__,                                                     \
+          __LINE__,                                                         \
+          assertion_caller_id,                                              \
+          blockIdx,                                                         \
+          threadIdx);                                                       \
+                                                                            \
+      assert(condition);                                                    \
```

Comment:
A design that evaluates `condition` once would call `assert(condition_result)` instead of `assert(condition)`, but `assert(condition_result)` would cause CUDA to print a less informative error message. That might be okay since our intention is to capture the error message anyway, but it ultimately reduces flexibility for users without much gain since, as you say, the assertions should (hopefully) be without side-effects.

> But perhaps you could call some internals like __assert_fail to provide an explicit message?

I'm not sure what you mean by this.

Reply:
`assert` is usually implemented as a macro somewhere inside CUDA. So you could depend on implementation detail and call the function that the macro calls underneath. In std c library it's usually `__assert_fail` (e.g. https://github.com/lattera/glibc/blob/master/assert/assert.h#L67)

Reply:
I've gone with __assert_fail.